lua/fidget/progress/handle.lua

```diff
@@ -0,0 +1,145 @@
+local notification = require("fidget.notification")
+
+local NEXT_ID = 0
+local PREFIX = "fidget-"
+
+local function next_token()
+  NEXT_ID = NEXT_ID + 1
+  return PREFIX .. NEXT_ID
+end
+
+--- @class ProgressHandle: ProgressMessage  A handle for a progress message, reactive to changes
+--- @field cancel fun(self: ProgressHandle) Cancel the task
+--- @field finish fun(self: ProgressHandle) Mark the task as complete
+--- @field report fun(self: ProgressHandle, props: { [string]: any } | ProgressMessage) Update one or more properties of the progress message
+
+local M = {}
+
+--- Create a new progress message, and return a handle to it for updating.
+--- The handle is a reactive object, so you can update its properties and the
+--- message will be updated accordingly. You can also use the `report` method to
+--- update multiple properties at once.
+---
+--- Example:
+---
+--- ```lua
+--- local progress = require("fidget.progress")
+---
+--- local handle = progress.create({
+---   title = "My Task",
+---   message = "Doing something...",
+---   lsp_name = "my_fake_lsp",
+---   percentage = 0,
+--- })
+---
+--- -- You can update properties directly and the
+--- -- progress message will be updated accordingly
+--- handle.message = "Doing something else..."
+---
+--- -- Or you can use the `report` method to bulk-update
+--- -- properties.
+--- handle:report({
+---   title = "The task status changed"
+---   message = "Doing another thing...",
+---   percentage = 50,
+--- })
+---
+--- -- You can also cancel the task (errors if not cancellable)
+--- handle:cancel()
+---
+--- -- Or mark it as complete (updates percentage to 100 automatically)
+--- handle:finish()
+--- ````
+---
+--- @param message { [string]: any} | ProgressMessage The initial progress message
+--- @return ProgressHandle
+function M.create(message)
+  local progress = require("fidget.progress")
+
+  message = message and vim.deepcopy(message) or {}
+
+  -- Generate a unique token for this message
+  message.token = next_token()
+
+  -- Set required fields
+  message.lsp_client = message.lsp_client or {
+    name = "fidget",
+  }
+  message.done = false
+
+  -- Cancellable by default
+  if message.cancellable == nil then
+    message.cancellable = true
+  end
+
+  -- Load the notification config
+  progress.load_config(message)
+
+  -- Initial update (for begin)
+  notification.notify(progress.format_progress(message))
+
+  -- Use a proxy with __gc to handle cleanup, ensuring that we don't
+  -- leak notifications if the user doesn't call finish() or cancel().
+  local proxy = newproxy(true)
+  getmetatable(proxy).__gc = function()
+    if not message.done then
+      message.done = true
+      notification.notify(progress.format_progress(message))
+    end
+  end
+
+  local handle = {}
+
+  setmetatable(handle, {
+    __newindex = function(_, k, v)
+      if k == "token" or k == "done" then
```

Comment:
What is the problem with setting `handle.done`? Semantically, `handle.done = true` could be treated the same as calling `handle:finish()`. I suppose it would be a bit odd to set `handle.done = false` after it is already `true`, but perhaps there's value in supporting "resurrecting" intermittent, long-running tasks.

The other problem here is that this does not forbid the user from, e.g., setting `handle.finish = bogus`. In fact, I think you end up doing that later (see comment below) when you set the methods after you set the metatable.

Reply:
You're right, that was probably overcomplicating things. I'll change that to only block `token` from being modified.